#17 Added recently vetted steps to use Data Hub with Grove

mariannemyers opened the pull request:
Steps are now in the docs to reflect the usage of mlModulesPath with a ui-modules directory added to the Data Hub code base.  I feel this is a cleaner implementation of Grove with the Data Hub.  If any more complexity is needed in to support the UI, like triggers or alerting, the split with two projects sharing a content database becomes unwieldy and require additional tooling/education on how to develop and deploy properly.  

This was vetted against DHF 4.3.1 and Data Hub 5.0.0.  It does require upgrading ml-gradle, even in 5.0, to use mlModulesPath.

Commit:
. #17 Added recently vetted steps to use Data Hub with Grove

withjam commented:
Needs updating for DHF 5

grtjn commented:
I think we discussed the usage of the word 'recommended'. Maybe we should phrase it something like: the approach that currently works best with both DHF and DHS is to migrate the few necessary artifacts into DHF gradle (option 1). The alternative approach is to maintain two gradle projects, one for DHF and documents database, the other for dedicated ui-modules, and ui app-server..

mariannemyers committed:
Update data-hub-framework.markdown

Minor edit as the newer Data Hub ml-gradle supports the mlModulesPaths property.

mariannemyers commented:
I've gotten a number of asks about this pattern recently.  Can we merge this in?

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -5,17 +5,110 @@ lead_text: ''
+[Optional] If you previously installed Grove to its default database via the `mlDeploy` command within the `marklogic` directory, remove the Grove-specific configs before proceeding:
+        
+    > cd grove/marklogic
+    > ./gradlew mlUndeploy
```
It is lacking a `-Pconfirm=true`, or would you like users to discover that themselves, as a means to make sure they know what they are doing? ;-)

grtjn replied:
just append " (and confirm)" or something alike..

mariannemyers replied:
Added the `-Pconfirm=true`

grtjn commented on _pages/guides/data-hub-framework.markdown:
````diff
@@ -5,17 +5,110 @@ lead_text: ''
+    dependencies {
+		  classpath "gradle.plugin.com.marklogic:ml-gradle:3.14.0"
+    }
+    ```
````
You could reword this slightly more generically, and tell everyone to check the ml-gradle version, saying it must be 3.14.0 or higher. (and a mention of 5.0.1 could be a side note "should only be necessary for dhf 5.0.1, or older dhf projects")

mariannemyers replied:
Updated for proper version requirements.

grtjn commented on _pages/guides/data-hub-framework.markdown:
````diff
@@ -5,17 +5,110 @@ lead_text: ''
+    ```
+
+9. Make any other necessary changes to the query options file. For example, you may need to remove the `<additional-query>` specifying that only docs in the `data` collection are returned.  There may be a Data Hub specific collection that is a natural fit to limit the search results for your Grove application.  
+
````
Consider adding a note to keep the additional-query in place, or you might start seeing es model defs and such in search results, which is typically not desirable.

grtjn replied:
We could suggest making a change along these lines:

```
      <cts:collection-query>
        <cts:uri>Entity1</cts:uri>
        <cts:uri>Entity2</cts:uri>
        .....
      </cts:collection-query>
```

grtjn replied:
With an added note that this prevents seeing the defs, and flows, etc..

mariannemyers replied:
Copied your fabulous example in!

grtjn commented on _pages/guides/data-hub-framework.markdown:
````diff
@@ -5,17 +5,110 @@ lead_text: ''
+    to 
+    
+    data-hub/src/main
+    ```
````
ml-content inside src/main is not deployed out of the box by ml-gradle afaik. This might require copying over the appropriate mlcp task from the grove build.gradle.

grtjn replied:
less required anyhow..

grtjn replied:
searched ml-gradle for ml-content, no hits..

grtjn replied:
On second review, there is built-in support for loading files, but it is called mlLoadData, and it looks by default for src/main/ml-data.. I suggest adding an instruction to rename the folder to `ui-data`, and adding the following line to `gradle.properties`: `mlDataPaths=src/main/ml-data,src/main/ui-data`

That seems to do the trick. mlLoadData is part of mlDeploy..

mariannemyers replied:
Added mlDataPaths and updated ui-data.

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -5,17 +5,110 @@ lead_text: ''
+There are some downsides to this approach:
+- UI specific code will reside in close proximity to backend specific configurations.  This might not be as preferable to some development teams.
+- Any CRUD operations performed by the Grove project will write to the Data Hub's FINAL database.
+
```
There is another 'down-side' (arguable). The security setup in Grove is more advanced and strict than you get with data-hub out of the box. The steps below skip over the ml-configs, and security in specific. I think we should at least add a note to encourage users to think carefully about the security needs for data processed in dhf/dhs.

grtjn replied:
'Lack' of security is comes from DHS that has its own approach to security. Not sure how to phrase that in here..

grtjn replied:
Security approach is different, best to drop the approach of grove, and utilize that of DHF/DHS..

mariannemyers replied:
Added a note per your comment.

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -5,17 +5,110 @@ lead_text: ''
+
+## Alternative Approach to Run Grove with the Data Hub
+
+In order to run a Grove UI project alongside the MarkLogic Data Hub Framework (DHF) project, we alternatively suggest creating a Grove project-specific app-server, which has its own modules database but points at the content database with the data you wish to visualize: most often the DHF project's FINAL database. The DHF project will be responsible for managing the content database (including se ...
```
Security will be a shared responsibility, where DHF has to give low-level access to data, and Grove can arrange higher-level access.

mariannemyers replied:
Added your comment.

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -5,17 +5,110 @@ lead_text: ''
+
+In order to run a Grove UI project alongside the MarkLogic Data Hub Framework (DHF) project, we alternatively suggest creating a Grove project-specific app-server, which has its own modules database but points at the content database with the data you wish to visualize: most often the DHF project's FINAL database. The DHF project will be responsible for managing the content database (including se ...
 
 Note that this implies some coordination between the DHF project and Grove. If possible, it is easiest to manage both projects within the same code repository. The Grove files should all be within a dedicated parent directory. (For example, `grove/ui`, `grove/middle-tier`, and `grove/marklogic`. Note that there is nothing magic about the `grove` directory name, so call it whatever you want.)
```
The other approach also requires coordination. Maybe this should be a global note at the top?

mariannemyers replied:
I have the same words in both areas of this document, so I think that will capture the coordination is required.

mariannemyers replied:
This comment did point out DHF is still referenced.  I updated the entire document to change DHF -> Data Hub.

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -5,17 +5,110 @@ lead_text: ''
+
+## Migrating Grove-related artifacts and code to the Data Hub
+
+In order to run a Grove UI project alongside a MarkLogic Data Hub, we recommend migrating Grove related artifacts and code from its code base into your Data Hub project.  
```
"This approach is tested against both DHF and DHS"

mariannemyers replied:
Added note.

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -54,21 +147,3 @@ There are some downsides to this approach:
 
 6. Run `./gradlew mlDeploy` from inside the `marklogic` directory. This command will deploy the new configuration, but it will not change the DHF project's content database, because you have removed all related configuration files.
```
The build.gradle likely needs to be edited, because setSchemaPermissions should be disabled (or removed altogether). DHF will be responsible of applying the correct doc permissions to the schemas (and triggers)

mariannemyers replied:
Steps are added to remove those parts.

grtjn commented:
Ping.. @mariannemyers can we test this against 5.1 as well? :)

And how far off are we to merging this?

grtjn commented on _pages/guides/data-hub-framework.markdown:
```diff
@@ -5,17 +5,110 @@ lead_text: ''
+1. Determine the MarkLogic host and port the data-hub-FINAL app server is running on.
+2. At the command line, from within the top level of your Grove project, (the directory with `middle-tier` and `ui` directories inside it), run `grove config` and enter the host and port of the data-hub-FINAL app server.  This will update environment variables for your project.
+3. Create the `ui-modules` directory inside the Data Hub project: `data-hub/src/main/ui-modules`.
+4. Update the Data Hub's `gradle.properties` to register this new location for MarkLogic module deployments.  Add the following line to `gradle.properites`:
```
typo 'properites'

grtjn commented:
@mariannemyers I think this is really close. Would be great if you could polish the text a little based on my comments, and update/merge the PR.. :)

mariannemyers committed:
Update for PR #17 findings.

mariannemyers committed:
Update sidenav for rename

mariannemyers committed:
Merge branch 'docs' into feature/GROVE-17
